Error when LHS of instanceof is Union of Primitives #18519

charlespierce opened the pull request:
<!--
Thank you for submitting a pull request!

Here's a checklist you might find useful.
[ ] There is an associated issue that is labelled
  'Bug' or 'Accepting PRs' or is in the Community milestone
[ ] Code is up-to-date with the `master` branch
[ ] You've successfully run `jake runtests` locally
[ ] You've signed the CLA
[ ] There are new or updated unit tests validating the change

Refer to CONTRIBUTING.MD for more details.
  https://github.com/Microsoft/TypeScript/blob/master/CONTRIBUTING.md
-->

Fixes #18519

Added additional check for if the left-hand operand of an `instanceof` is a union of primitives and raise an error in that case as well.

mhegazy commented:
@DanielRosenwasser can you review this change.

charlespierce commented:
@DanielRosenwasser @mhegazy Just wanted to check if there were any issues with this PR, so I can make any needed changes before it goes stale.

DanielRosenwasser commented on src/compiler/checker.ts:
```diff
@@ -17604,7 +17604,9 @@ namespace ts {
             // NOTE: do not raise error if leftType is unknown as related error was already reported
-            if (!isTypeAny(leftType) && isTypeAssignableToKind(leftType, TypeFlags.Primitive)) {
+            if (!isTypeAny(leftType) &&
+                (isTypeAssignableToKind(leftType, TypeFlags.Primitive) ||
```
I wonder if it makes sense to hae `isTypeAssignableToKind` just work recursively for all constituents if given a union type. Perhaps make an `allTypesAssignableToKind` helper below it to keep the change scoped.

Otherwise, this seems fine.

charlespierce commented:
@DanielRosenwasser I like keeping the change scoped, since `isTypeAssignableToKind` is used in a lot of places and there may be some unintended consequences of adding support for union types to it. Updated to use an `allTypesAssignableToKind` helper.

Also updated the test to include the case `(number | string) | number`, which would have failed in the previous implementation since it wasn't recursively checking for union types.

mhegazy commented on src/compiler/checker.ts:
```diff
@@ -17587,6 +17587,16 @@ namespace ts {
         }
 
+        function allTypesAssignableToKind(source: Type, kind: TypeFlags, strict?: boolean): boolean {
+            if (source.flags & TypeFlags.Union) {
```
```ts
return source.flags & TypeFlags.Union ?  every((source as UnionType).types, subType => allTypesAssignableToKind(subType, kind, strict)) : isTypeAssignableToKind(source, kind, strict);
```

charlespierce replied:
Updated

charlespierce commented:
@mhegazy @DanielRosenwasser Any additional concerns?

mhegazy commented:
Sorry for the delay.
